rstream/client.py

```diff
@@ -270,18 +291,17 @@ def is_started(self) -> bool:
     async def close(self) -> None:
         logger.info("Stopping client %s:%s", self.host, self.port)
 
-        if self._conn is None:
-            return
+        if self._conn is not None:
 
-        if self.is_started:
-            await self.sync_request(
-                schema.Close(
-                    self._corr_id_seq.next(),
-                    code=1,
-                    reason="OK",
-                ),
-                resp_schema=schema.CloseResponse,
-            )
+            if self.is_started:
+                await self.sync_request(
+                    schema.Close(
+                        self._corr_id_seq.next(),
+                        code=1,
+                        reason="OK",
+                    ),
+                    resp_schema=schema.CloseResponse,
+                )
```

Comment:
@garvenlee yeah probaly it would be safer to put a timeout in the schema.Close here.

The task can be cancelled in this case. In stream messages are kept in the server and they can be consumed again. (I think also other clients work in the same way adding @Gsantomaggio to confirm)

Reply:
HI @garvenlee I put a modification with a timeout on sync_request of schema Close().

Regarding the second point to avoid further complications, we could document to not call producer.close() inside the callback but to call it in the main flow.
For example the callback could just notify that there was a connection issue and then the producer.close() could be called on the main flow.

Reply:
point 1: You can add an arg `timeout` to `sync_request`, and reuse `wait_frame` to apply timeout.

point 2: I think it’s worth wrapping `connection_closed_handler` as a task in `listener`. It has little effect on other logic, but helps to solve the above scenario. And this makes `Client.close` more straightforward.

```Python
async def close(self):
    …

    tasks = []
    for task in self.tasks.values():
        task.cancel()
        tasks.append(task)
    await asyncio.wait(tasks)
    
    await self._conn.close()
    self._conn = None
    …


```